src/Adapters/TraceListener.Shared/ApplicationInsightsTraceListener.cs

```diff
@@ -47,6 +47,16 @@ public ApplicationInsightsTraceListener(string instrumentationKey)
             this.TelemetryClient.Context.GetInternalContext().SdkVersion = SdkVersionUtils.GetSdkVersion("sd:");
         }
 
+        /// <summary>
+        /// Initializes a new instance of the ApplicationInsightsTraceListener class, specifying a TelemetryClient to use
+        /// </summary>
+        /// <param name="telemetryClient">TelemetryClient to use for tracing</param>
+        public ApplicationInsightsTraceListener(TelemetryClient telemetryClient)
```

Comment:
Typical way doing it will be to code it like this:

``` csharp
class SessionIdTelemetryInitializer : ITelemetryInitializer
{
  public void Process(ITelemetry t)
  {
    t.Context.Session.Id = GetSessionIdForCurrentThread();
  }
}

var config = new TelemetryConfiguraiton();
config.TelemetryInitializers.Add(new SessionIdTelemetryInitializer());

var listener = new ApplicationInsightsTraceListener(config);
```

This way you can control session ID dynamically and do not need to synchronize threads.

Will this satisfy your scenario?